[jaspertzx] iP

Jaspertzx opened the pull request:
### BobBot


> "Soon is not as good as now."  - Seth Godin ([source](https://dansilvestre.com/productivity-quotes/#9_%E2%80%9CSoon_is_not_as_good_as_now%E2%80%9D_%E2%80%93_Seth_Godin))

BobBot helps to free your memory space by taking down things you need to do. It's,

- command-line interface (so super fast)
- easy to learn
- ~~super duper extra duper~~ really fast to use

All you really have to do to start os:

1. Ensure you have Java11 installed.
2. Download it from [here](https://github.com/Jaspertzx/ip/releases).
3. Double-click it.
4. Add your task!
5. Let it manage it for you!

And it is **_FREE_**! 😀

Features

- [ ] Manageing tasks
- [ ] Managing Deadlines (Coming soon!)
- [ ] Reminders (Coming soon!)

***
If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Jaspertzx committed:
Changed name of bot and included more System prints

Jaspertzx committed:
added reader, and some personality!

Jaspertzx committed:
begun storing commands and listing them out with list command

Jaspertzx committed:
Added new Task Class, and added mark and ummark

Jaspertzx committed:
added 3 subclasses of task

Jaspertzx committed:
edited the test files and confirmed that there is no error

Jaspertzx committed:
Added in exceptions and did some error handling throughout Duke.java

Jaspertzx committed:
added comments, and delete function

Jaspertzx committed:
Add automatic task saving and loading feature

Tasks are not persistent between sessions.

Persistent storage allows users to pick up where they left off.

Let's introduce saving & loading to `./data/duke.txt`.
Use OS-independent file handling to ensure compatibility.
Create reader & writer classes to abstract out the IO responsibility.

This approach ensures data durability and enhances the chatbot's usability.
It supports task management across sessions without user intervention.

Jaspertzx committed:
Merge branch-Level-7 into master

Jaspertzx committed:
Add date and time parsing for task deadlines

Chatbot treats deadline commands as plain strings.

Add parsing of strings to convert them to LocalDateTime obj,
as well as error handling for wrongly formatted inputs.

This approach is taken for clarity and ease of use, and
allows for future use of the java.time API for timezone
related operations.

Jaspertzx committed:
Merge branch-Level-8 into master

Jaspertzx committed:
Refactor code for better OOP design

The application is messy and scattered.

Makes it hard to maintain and extend.
Also makes it difficult to understand code.

Refactor application to adopt more OOP.
Creat new parser and dependencies to abstract
from the main application.

To improve code maintainability and reability,
allow for easy edits in future development.

Jaspertzx committed:
Merge branch-A-MoreOOP into master

Jaspertzx committed:
Create packages for classes

All classes were located in the default package.

Issues with messy structure and possible name
conflicts. Also made it difficult to navigate
the codebase.

Create multiple packages and shifted smiliar or
related classes together within these packages.

Allows for easier traversal through the coding
logic, and grouping of similar logic classes.

Jaspertzx committed:
Create packages for classes

All classes were located in the default package.

Issues with messy structure and possible name
conflicts. Also made it difficult to navigate
the codebase.

Create multiple packages and shifted smiliar or
related classes together within these packages.

Allows for easier traversal through the coding
logic, and grouping of similar logic classes.

Jaspertzx committed:
Merge branch-A-Packages into master

Jaspertzx committed:
Merge remote-tracking branch 'main-repo/add-gradle-support' into branch-A-Gradle

Jaspertzx committed:
Add and automate build with Gradle

Project build are being managed manually.
Have to manually compile all the java file.

Inefficient and slow, manual process as well.
Could be running outdated .class files if
not done properly.

Add Gradle from master branch by merge.

Gradle helps with setting up the Duke application
and helps to automate the manual tasks.

Jaspertzx committed:
Merge branch-A-Gradle into master

Jaspertzx committed:
Add JUnit Tests to application

Testing is very manual, by printing out
variable and confirming each input by
hand.

This is not scalable and not practical
when the codebase becomes much bigger
and there are more dependencies in the
future.

Add JUnit tests for certain classes
based on importance. To ensure that the
expected behaviour is as observed and
does not change with further development.

JUnit tests are great as it is integrated
with gradle, thus becoming easy to maintain
and run.

Jaspertzx committed:
Merge branch 'branch-A-JUnit'

Jaspertzx committed:
Package main application into JAR file

The application needs to be compiled and
ran with java from the terminal or IDE.

This will make the application difficult
more difficult to run for end users.

Package main application into java using
shadowjar, allow us to run the application
with a simple java - jar <filename>.

This will allow for an easier time to
test and use the application.

Jaspertzx committed:
Merge branch-A-Jar into master

Jaspertzx committed:
Add JavaDoc to application

Not all classes have completed Javadoc comments.

This makes reading the code confusing and time-
consuming as you have to understand the code before
making sense of it's purpose.

Add JavaDoc to each class and most methods.

This improves code readability and maintainability
as it allows readers to quickly understand what
the class and methods are used for.

Jaspertzx committed:
Merge branch-A-JavaDoc into master

Jaspertzx committed:
Adhere to coding standard

The Java codes in the application do not
adhere to the coding standard.

This may result in the code being hard
to read or understand, and therefore even
harder to maintain

Change codes to adhere to coding standard,
imported checkstyler and used Gradle to
run.

Allow gradle to check Java code for
readability and makes code more readable.

Jaspertzx committed:
Merge branch-A-CodingStandard into master

Jaspertzx committed:
Add find command to Chatbot application

There is no find command in the application,
users have to list all task and search manually.

This is no intuitive and causes the user to
manually search for his/her task.

Add find command to filter task for the user.

Allows user to quickly find the task they are
looking for.

Jaspertzx committed:
merge branch-Level-9 into master

ChocoRagdoll commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,44 @@
+
+import parser.ParseExecutionable;
```
I like how you explicitly listed out import statements.

ChocoRagdoll commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,191 @@
+package parser;
+
+import java.time.LocalDateTime;
```
I like how you have a consistent ordering of import statements.

ChocoRagdoll commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,191 @@
+**/
+public class Parser {
+
+    static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
I like how this constant is correctly named in all uppercase.

ChocoRagdoll commented on src/main/java/util/Messages.java:
```diff
@@ -0,0 +1,25 @@
+ * The Message class provides a centralized repository of strings used throughout
+ * the application to display various forms of messages.
+ */
+public class Messages {
```
Should we make the field private and integrate this class into the Ui class?

ChocoRagdoll commented on src/main/java/task/ActionTask.java:
```diff
@@ -0,0 +1,131 @@
+    @Override
+    public String execute(TaskStorage taskStorage) {
+        String printMessage = "";
+        switch (this.action) {
```
I like how the coding standard is closely followed.

ChocoRagdoll commented on src/main/java/task/TaskStorage.java:
```diff
@@ -0,0 +1,140 @@
+ * such as addition, deletion, and editing Tasks.
+ */
+public class TaskStorage {
+    private ArrayList<Task> sl;
```
Should this variable be given a longer name?

ChocoRagdoll commented on src/test/java/DeadlineTest.java:
```diff
@@ -0,0 +1,30 @@
+public class DeadlineTest {
+
+    @Test
+    public void deadlineConstructTest() {
```
An alternative way of naming these tests is `featureUnderTest_testScenario_expectedBehavior()`, although this should not be compulsory.

ChocoRagdoll reviewed (commented):
LGTM 😸
Just a few nits to fix.

Jaspertzx committed:
Add CheckStyle to check for consistency

Codebase lacks the enforcement of proper
coding standards.

This could lead to messy, unformatted
code that makes the Codebase hard
to read and maintain.

Installed and configured a CheckStyle
xml file to run with gradlew, such
that every time I run the project,
it checks for the Java coding style.

This will ensure that before commiting
each version, I will have an
opportunity to check the Java style
before pushing to the repo.

Jaspertzx committed:
Merge branch-A-CheckStyle into master

Jaspertzx committed:
Add Graphical User Interface to app

The chatbot uses the terminal to communicate.

This might not be practical with non-tech
savvy users, and the terminal isn't great
to look/use sometimes.

Add a GUI using JavaFX to provide for users
to visually look at their input and output.

This makes it more intuitive that it is a
Chatbot, and maybe easier for certain users.

Jaspertzx committed:
Merge branch-Level-10 into master.

Jaspertzx committed:
Add Graphical User Interface to app

The chatbot uses the terminal to communicate.

This might not be practical with non-tech
savvy users, and the terminal isn't great
to look/use sometimes.

Add a GUI using JavaFX to provide for users
to visually look at their input and output.
Add JavaFX FXML files and their respective
controllers to help abstract out elements
in the GUI for easier readability and
maintainability.

This makes it more intuitive that it is a
Chatbot, and maybe easier for certain users.

Jaspertzx committed:
Merge branch-Level-10 into master

Jaspertzx committed:
Implement Assertions for validation

The code does not have any internal validation.
This may allow errors during development that
may go unnotice, and may make the code unreliable.

This may make the code unreliable as sometimes
external tests such as JUnit testing, not not
be able to create all possible scenarios that
may happen during production.

Add assertions to classes with in the codebase.

This will allow me to internally test our codebase
before going to to run JUnit for external testing.
This allows my testing to be comprehensive and to
catch more unexpected errors that we may not have
thought about during my development.

Jaspertzx committed:
Merge pull request #1 from Jaspertzx/branch-A-Assertions

Implement Assertions for validation

Jaspertzx committed:
Revert "Implement Assertions for validation"

Jaspertzx committed:
Merge pull request #2 from Jaspertzx/revert-1-branch-A-Assertions

Revert "Implement Assertions for validation"

Jaspertzx committed:
Ensure code quality in codebase

The code base has duplicated code, unclear naming
conventions. This makes it unnecessarily long and
complex.

This makes the code base hard to maintain and edit
in future iterations, and harder for other
developers to read and implement their logic as
well.

Refractor and fix naming conventions across the
java files. Additionally, removed repeated code
from the main UI java files to shorten the lines
of code.

This allows for better readability
for other reads, making the code more concise and
easier to understand as well.

Jaspertzx committed:
Merge branch 'master' into branch-A-CodeQuality

Jaspertzx committed:
Merge pull request #3 from Jaspertzx/branch-A-CodeQuality

Ensure code quality in codebase

Jaspertzx committed:
Revert "Ensure code quality in codebase"

Jaspertzx committed:
Merge pull request #4 from Jaspertzx/revert-3-branch-A-CodeQuality

Revert "Ensure code quality in codebase"

Jaspertzx committed:
Implement Assertion and fixing errors in code

The current code has some styling errors in the
code.

This may cause the code to be difficult to read
or misleading.

Adjust the naming of functions and variables to
the Java convention.

This will allow the code to be easily read.

Jaspertzx committed:
Ensure code quality and fix errors in codebase.

The codebase has some naming style warnings, and
bad variables names such as c or n.

These do not provide enough detail for the other
readers, which can cause it to be misleading.

Edit the naming violations to ensure that they
are compliant with the convention.

This will increase code reability and allow
my readers to have a better understanding of
the code.

Jaspertzx committed:
Merge branch 'master' into branch-A-Assertions

Jaspertzx committed:
Merge pull request #5 from Jaspertzx/branch-A-Assertions

Implement Assertion and fixing errors in code

Jaspertzx committed:
Resolve merge conflicts

Jaspertzx committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/Jaspertzx/ip into branch-A-CodeQuality

Jaspertzx committed:
Merge pull request #6 from Jaspertzx/branch-A-CodeQuality

Ensure code quality in codebase

Jaspertzx committed:
Utilize Java Enums for different types

The ActionTask uses strings to represent different
types of actions. This allows it to handle a fixed
set of variables.

This could lead to errors and makes the code hard
to implement, as developers have to read and match
the words for the corresponding actions.

Let's change the String list into an Enum, and use
it to indicate the respective action type of
ActionTask.

This will allow only defined types to be created,
enforcing proper type security and ensuring that
it is easy to define and debug.

Jaspertzx committed:
Merge pull request #7 from Jaspertzx/branch-A-Enums

Utilize Java Enums for different types

Jaspertzx committed:
Add Continuous Integration

The codebase lacks autonomous testing and
deployment of its code.

This may affect the reliability of the code,
due to infrequent testing as I may forget.
Additionally, automating this process ensures
high code quality in every commit.

Let's add the .yml configuration file for
github to use to configure the Continuous
Integration.

This will allow github actions to help test and
deploy the code as it is directly integrated
into my github repository.

I referenced the following materials:
se-education.org/guides/tutorials/githubActions.html

Jaspertzx committed:
Merge and Add Continuous Integration

The codebase and repo do not have Continous 
Integration set up.

This may result in the code being unreliable,
as it is possible to push a code that cannot
be built using gradlew.

Let's add the .yml configuration file for
github to use to configure the Continuous
Integration. Futhermore, we update our 
Github PAT to ensure that we can make
the necessary changes to workflows.

This will allow GitHub actions to help test and
deploy the code as it is directly integrated
into my GitHub repository.

I referenced the following materials:
se-education.org/guides/tutorials/githubActions.html

Jaspertzx committed:
Add sorting to Storage

The code does not sort Tasks according to their
deadlines or start date.

This may lead users to miss their deadlines as
they are not chronologically ordered.

Implement comparator to each of the Tasks's
subclass.

This allows us to sort the Tasks according to
their urgency, if applicable, and name.

Jaspertzx committed:
Merge pull request #10

The code has no ability to sort the Tasks
for the user to see it chronologically.

This is not intuitive and the user may
miss important deadlines due to this.

Implement a Comparator on each 
Task and override it's compareTo
method.

This will allow me to sort the storage
list by the respective Tasks, allowing
users to see their Tasks in a
chronological order.

Jaspertzx committed:
Add User UserGuide

Current code base has no UserGuide created, on
GitHub under docs.

This may cause users to use the application
without clear instructions, thus not allowing
them to effectively use the features of BobBot.

Add comprehensive user guide to the docs folder,
for GitHub and users to refer to.

This will allow users to easily refer to the guide
created by the sole developer of BobBot, allowing
them to effectively use the application.

Referred to the links provided for this branch.

Jaspertzx committed:
Merge pull request #11

Current code base does not have proper 
UserGuide created, on GitHub under docs.

This may cause users to use the application
without clear instructions, thus not allowing
them to effectively use the features of BobBot.

Add comprehensive user guide to the docs folder,
for GitHub and users to refer to.

This will allow users to easily refer to the guide 
created by the sole developer of BobBot, allowing
them to effectively use the application.

Referred to the links provided for this branch.

Jaspertzx committed:
Add JUnit testing and fix small errors

There are not enough Junit test files to
comprehensively test all functions that I have
created.

This may make the code unreliable as future
changes may break the codebase that would not be
easy to spot.

Add more JUnit testing, and some minor bug fixes
for the application that I found while JUnit test.

This will ensure that the code remains reliable
and it is easier for me to check my code before
committing.

No further references.

Jaspertzx committed:
Fix readme and build file

The code is working great, but is having issues
in the userguide. This issue would be either
grammar errors or unclear examples.

This may cause the user to have issues
reading the userguide, and may see unnecessary
comments in the build file.

Edit the readme to better guide users on the use
of the application. Remove unnecessary comments
in the build file too.

This will allow the user a better experience
following the guide and code, without any
unnecessary comments.

No further references.

Jaspertzx committed:
Fix Linux compatability with app

The application has issues being ran on the
Linux operating system. This incluides Ubuntu,
Kali Linux, etc.

This affects users as linux users are unable to
run the BobBot Application, and that makes the
application have a smaller target audience.

Fix the build gradle file to include the linux
configurtations to utilize the respective
dependency.

This will allow linux users to use the app,
opening up the range of audience.

No further references.
